Feature quick player info access

Undid-Iridium opened the pull request:
[Potentially breaking change]

Instead of passing back IEnumerables after having to iterate every player again and again, instead use a dictionary that updates on  player role change, and when a player leaves. That's the whole point - I am tired but from my testing it works fine. Feedback is appreciated. Thanks (Also for the dictionary add/remove, yes I could assign every possible role to a new dictionary but I was lazy and this at least allows dynamic updates in case someone messes up and does not update it)

Also it seems dev and 5.3 have some different changes, so you'll see some weird changes that were not my own, but rather the merge differences

Undid-Iridium commented on Exiled.API/Features/Player.cs:
```diff
@@ -1017,21 +1031,33 @@ public bool IsSpawnProtected
-        public static IEnumerable<Player> Get(RoleType role) => List.Where(player => player.Role == role);
+        public static IEnumerable<Player> Get(RoleType role)
+        {
+            InstantiatedRolesToPlayers.TryGetValue(role, out HashSet<Player> playersForRole);
```
I don't know if it's better to just do a nullable check or do a TryGetValue here ¯\_(ツ)_/¯ - brain kinda fried

IRacle1 replied:
specifically here `TryGetValue` is better, imho

Undid-Iridium commented on Exiled.Events/Patches/Events/Player/Spawning.cs:
```diff
@@ -34,11 +34,14 @@ private static IEnumerable<CodeInstruction> Transpiler(IEnumerable<CodeInstructi
             // Remove all existing this._pms.OnPlayerClassChange calls (we will want to call this ourselves after our even fires, to allow their spawn position to change.)
-            newInstructions.RemoveAll(i =>
-                i.opcode == OpCodes.Call && (MethodInfo)i.operand == AccessTools.Method(typeof(PlayerMovementSync), nameof(PlayerMovementSync.OnPlayerClassChange)));
+            foreach (CodeInstruction instruction in newInstructions.FindAll(i =>
```
Uh, 5.3 branch seems to have gone backwards on this, bad?

IRacle1 replied:
There are a lot of changes that do not apply to your PR, and are in other branches in Exiled, yea, bad, revert all these changes

IRacle1 commented on Exiled.API/Features/Player.cs:
```diff
@@ -122,6 +122,21 @@ public Player(GameObject gameObject)
+        /// <summary>
+        /// Gets all the current players of a specific role.
+        /// </summary>
+        public static Dictionary<RoleType, HashSet<Player>> InstantiatedRolesToPlayers { get; internal set; } = new();
```
First, I dont think that's those dicts are really needed.
To me, it's just an inefficient memory usage, LINQ in `Player.Get` better as for me.

Undid-Iridium replied:
I do agree storing HashSet<Player> is a little expensive; however, speed wise of access has to be significantly better than LINQ foreach every single time you want a specific player of Role, Side, and Team. However, it's up to the feedback I get to decide if this is worth the time (AKA you, Nao, Build, etc)

Undid-Iridium commented on Exiled.Events/Handlers/Player.cs:
```diff
@@ -534,7 +535,10 @@ public static class Player
-        public static void OnVerified(VerifiedEventArgs ev) => Verified.InvokeSafely(ev);
+        public static void OnVerified(VerifiedEventArgs ev)
+        {
+            Verified.InvokeSafely(ev);
```
Sorry, I ran into a bug I could not figure out and was testing OnLeft and OnVerified - I reverted OnLeft but forgot OnVerified (it was 4 am :( )

Undid-Iridium commented on Exiled.Events/Patches/Events/Player/Destroying.cs:
```diff
@@ -41,8 +41,6 @@ private static IEnumerable<CodeInstruction> Transpiler(IEnumerable<CodeInstructi
                 new(OpCodes.Call, Method(typeof(Player), nameof(Player.Get), new[] { typeof(ReferenceHub) })),
                 new(OpCodes.Stloc_S, player.LocalIndex),
                 new(OpCodes.Ldloc_S, player.LocalIndex),
-                new(OpCodes.Ldnull),
```
Done, thanks!

Undid-Iridium commented:
**Do not merge until consensus is agreed upon**

louis1706 commented on Exiled.Events/EventArgs/ChangingRoleEventArgs.cs:
```diff
@@ -75,5 +78,53 @@ public ChangingRoleEventArgs(Player player, RoleType newRole, bool shouldPreserv
+            }
+            else
+            {
+                Player.InstantiatedRolesToPlayers.Add(NewRole, new HashSet<Player>() { Player });
```
Create all this value when server start with empty  hashset and so you will not need to make this method like an yandere dev :3

Undid-Iridium replied:
This behavior prevents developer mistakes. I already considered allocating all roles, side, team preemptively; however, if a developer messes it up or someone misses a new role, side, or team - this code will not run into a problem as it dynamic. I would make this change if it is a consensus choice

louis1706 replied:
Developer should don't have access to this things just make it private

Undid-Iridium replied:
Hmm fair point, I agree then - I'll make change tomorrow,  Thanks!

louis1706 commented on Exiled.Events/Patches/Events/Player/Left.cs:
```diff
@@ -73,10 +73,34 @@ private static IEnumerable<CodeInstruction> Transpiler(IEnumerable<CodeInstructi
+                previousTeamPlayers.Remove(player);
+            }
+
+            if (Player.InstantiatedSideToPlayers.TryGetValue(player.Role.Side, out HashSet<Player> previousSidePlayers))
```
For here i would say to use Player.InstantiatedToPlayers[player.Role.Side].Remove(player)

Undid-Iridium replied:
Not guaranteed to be there; hence; previous comment explains why there is a protection here

louis1706 replied:
So protection will be not needed ^^

Undid-Iridium replied:
Done, thanks!

louis1706 replied:
Not done everywhere

Undid-Iridium replied:
Done

Undid-Iridium commented on Exiled.Events/Patches/Events/Player/SearchingPickupEvent.cs:
```diff
@@ -82,7 +82,7 @@ private static IEnumerable<CodeInstruction> Transpiler(IEnumerable<CodeInstructi
                 new(OpCodes.Ldarg_1),
                 new(OpCodes.Initobj, typeof(SearchSession)),
-                new(OpCodes.Ldarg_2),
+                new(OpCodes.Ldarga_S, 2),
```
```
Previous
new(OpCodes.Ldarg_2), <-- Load variable (not an address)
new(OpCodes.Ldnull), <-- Load null
new(OpCodes.Stind_Ref), <-- Store null in address
```

```
Current
new(OpCodes.Ldarga_S, 2), <-- Load variable address
new(OpCodes.Ldnull), <-- Load null
new(OpCodes.Stind_Ref), <-- Store null in address
```

Undid-Iridium replied:
Done, Thanks!

Undid-Iridium commented on Exiled.Events/EventArgs/ChangingRoleEventArgs.cs:
```diff
@@ -39,6 +40,8 @@ public ChangingRoleEventArgs(Player player, RoleType newRole, bool shouldPreserv
             Lite = shouldPreservePosition;
             Reason = (SpawnReason)reason;
+
+            ConsolidatePlayerInformation();
```
Done, Thanks!

IRacle1 commented on Exiled.Events/Patches/Events/Player/ChangingRole.cs:
```diff
@@ -76,11 +76,14 @@ private static IEnumerable<CodeInstruction> Transpiler(IEnumerable<CodeInstructi
                 // Handlers.Player.OnChangingRole(ev)
                 new(OpCodes.Call, Method(typeof(Player), nameof(Player.OnChangingRole))),
 
+                new(OpCodes.Callvirt, Method(typeof(ChangingRoleEventArgs), nameof(ChangingRoleEventArgs.ConsolidatePlayerInformation))),
+
                 // if (!ev.IsAllowed)
                 //    return;
                 new(OpCodes.Callvirt, PropertyGetter(typeof(ChangingRoleEventArgs), nameof(ChangingRoleEventArgs.IsAllowed))),
```
🙂 now it will be broke when `ev.IsAllowed = false`

Undid-Iridium replied:
If the event is false their role does not change

IRacle1 replied:
Yes, I overreacted, but logically it would be correct to put it after the IsAllowed check(You've already done it.)

IRacle1 commented on Exiled.API/Features/Player.cs:
```diff
@@ -1016,21 +1071,33 @@ public bool IsSpawnProtected
         /// <param name="side">The players' side.</param>
         /// <returns>The filtered <see cref="IEnumerable{T}"/>.</returns>
-        public static IEnumerable<Player> Get(Side side) => List.Where(player => player.Role.Side == side);
+        public static IEnumerable<Player> Get(Side side)
```
Yes, it was better until you set the default values in the dictionaries

Undid-Iridium replied:
Done, Thanks!

Undid-Iridium commented on Exiled.Events/Patches/Events/Player/ChangingRole.cs:
```diff
@@ -82,13 +82,14 @@ private static IEnumerable<CodeInstruction> Transpiler(IEnumerable<CodeInstructi
                 new(OpCodes.Callvirt, PropertyGetter(typeof(ChangingRoleEventArgs), nameof(ChangingRoleEventArgs.IsAllowed))),
                 new(OpCodes.Brfalse_S, returnLabel),
 
+                // PlayerRoles being updated.
+                new(OpCodes.Callvirt, Method(typeof(ChangingRoleEventArgs), nameof(ChangingRoleEventArgs.ConsolidatePlayerInformation))),
+
```
Check again

Undid-Iridium replied:
![image](https://user-images.githubusercontent.com/24619207/176468997-a5e76b9b-91bc-48a9-907d-6183a4baf6ad.png)

??? git ???

Undid-Iridium replied:
![image](https://user-images.githubusercontent.com/24619207/176469043-f6a1fbdc-c600-4a14-8035-5dc7c3ac4796.png)

IRacle1 replied:
😈

Undid-Iridium replied:
I don't know how that happened but OK - sorry I guess lmao

Undid-Iridium replied:
I did, I think github was freaking out because I kept pushing changes every minutes, I kept refreshing and it didn't show my previous commit until a few minutes later, way after you make this review. Very weird

Undid-Iridium replied:
Also I've had 8 hours of sleep in the last two days, my brain a little fried so thanks for the help @louis1706 @IRacle1

IRacle1 replied:
don't worry.

IRacle1 replied:
But the problem with git is really weird, I've never had that

louis1706 commented on Exiled.API/Features/Player.cs:
```diff
@@ -1016,21 +1071,21 @@ public bool IsSpawnProtected
         /// <param name="side">The players' side.</param>
         /// <returns>The filtered <see cref="IEnumerable{T}"/>.</returns>
-        public static IEnumerable<Player> Get(Side side) => List.Where(player => player.Role.Side == side);
+        public static IEnumerable<Player> Get(Side side) => InstantiatedSideToPlayers?[side];
 
         /// <summary>
         /// Gets a <see cref="Player"/> <see cref="IEnumerable{T}"/> filtered by team. Can be empty.
         /// </summary>
```
Public static IEnumerable<Player> Get(Side side) => InstantiatedSideToPlayers[side]

louis1706 commented on Exiled.API/Features/Player.cs:
```diff
@@ -1016,21 +1071,21 @@ public bool IsSpawnProtected
         /// <param name="role">The players' role.</param>
         /// <returns>The filtered <see cref="IEnumerable{T}"/>.</returns>
-        public static IEnumerable<Player> Get(RoleType role) => List.Where(player => player.Role == role);
+        public static IEnumerable<Player> Get(RoleType role) => InstantiatedRolesToPlayers?[role];
```
Why you use "?" It's should not being null. if it's null it's because the server have not finish to start and the dev have done a skillissue because calling Player.Get before server have finish to start is pretty useless

Undid-Iridium replied:
Fine.. :kek

moddedmcplayer commented:
Way to prevent Player from being added (npcs) and option to reset & resync lists incase something happens / on round end?

louis1706 commented:
> Way to prevent Player from being added (npcs) and option to reset & resync lists incase something happens / on round end?

Npc should not being in Player.List

Undid-Iridium commented:
> Way to prevent Player from being added (npcs) and option to reset & resync lists incase something happens / on round end?

So, I got 4 hours of sleep again, because I took a nap like an idiot, and like a bigger idiot I forgot to set an alarm so, forgive me for my stupidity (I promise I will sleep correctly today)

So you want an NPC to not be added to a list of players by role - I cannot think of a scenario where someone would want to do that, but my previous logic would have allows you to modify it after a player changed role. If you can give me a good scenario I will add it (Not trying to be rude I just can't think very well right now)

For the second half of your question, on server restart, everything gets wiped, no? However, I am not opposed to adding that functionality - Would you prefer a static api call for a resync, and a EventHander extra field to reset on round end?

moddedmcplayer commented:
Yeah that would Work fine, also funny npcs API in exiled 6 (hopefully)

Undid-Iridium commented:
@moddedmcplayer I have added what you've asked - see ReSyncAllInstantiatedContainers, ResetAllInstantiatedContainers, and ReSyncContainersToDefault. @IRacle1 Can you confirm my understanding of new Dictionary(previous dictionary) - This process is a shallow copy right, it won't copy the underlying reference, correct? (I added internal default dictionaries for this system)

IRacle1 commented:
> @moddedmcplayer I have added what you've asked - see ReSyncAllInstantiatedContainers, ResetAllInstantiatedContainers, and ReSyncContainersToDefault. @IRacle1 Can you confirm my understanding of new Dictionary(previous dictionary) - This process is a shallow copy right, it won't copy the underlying reference, correct? (I added internal default dictionaries for this system)

yea, rigth

Undid-Iridium commented:
> > @moddedmcplayer I have added what you've asked - see ReSyncAllInstantiatedContainers, ResetAllInstantiatedContainers, and ReSyncContainersToDefault. @IRacle1 Can you confirm my understanding of new Dictionary(previous dictionary) - This process is a shallow copy right, it won't copy the underlying reference, correct? (I added internal default dictionaries for this system)
> 
> yea, truth

Ty ❤️

IRacle1 commented:
But I'm not sure that the links to HashSet's will change (more precisely, I'm 90% sure that the HashSet's will be copied into the dictionaries by ref)

IRacle1 commented on Exiled.API/Features/Player.cs:
```diff
@@ -1006,31 +1006,142 @@ public bool IsSpawnProtected
+        /// </summary>
+        public static void ReSyncContainersToDefault()
+        {
+            InstantiatedRolesToPlayers = new Dictionary<RoleType, HashSet<Player>>(InstantiatedRolesToPlayersInternal);
```
use `new()` without type

Undid-Iridium replied:
See previous comment

IRacle1 replied:
🙂

IRacle1 commented on Exiled.Events/Handlers/Player.cs:
```diff
@@ -8,6 +8,7 @@
 namespace Exiled.Events.Handlers
 {
     using System;
+    using System.Collections.Generic;
```
why

Undid-Iridium replied:
Not sure what you mean, you need using

IRacle1 replied:
Yeah, it's not needed here.

Undid-Iridium replied:
![image](https://user-images.githubusercontent.com/24619207/177051251-f1f53bd8-3e9b-47d8-b343-9419c62a2eb9.png)
Hmm are you sure?

IRacle1 commented on Exiled.Events/EventArgs/RoundEndedEventArgs.cs:
```diff
@@ -43,5 +44,13 @@ public RoundEndedEventArgs(LeadingTeam leadingTeam, RoundSummary.SumInfo_ClassLi
+        /// <summary>
+        /// Resync's player information in Player class.
+        /// </summary>
+        public void ReSyncPlayerInformation()
```
Why?

Undid-Iridium replied:
Oh, for modded

Undid-Iridium replied:
![image](https://user-images.githubusercontent.com/24619207/177050703-0d62528f-84bc-41fc-889c-39bbc2a8c36b.png)

Undid-Iridium replied:
Technically, it's a wrapper just for people to not get confused, but they can also just do Player.ReSyncAllInstantiatedContainers(); so I can remove it but I offered it so I did it

IRacle1 replied:
As for me it makes no sense, the round will exist for some time after the call of this event, and the players still have roles(logically)

Undid-Iridium replied:
I mean, I agree. I made that point earlier that the data will be reset anyway on server restart - If you want me to remove it I can, I only added it as a helpful call per Modded's request

Undid-Iridium commented on Exiled.API/Features/Player.cs:
```diff
@@ -1006,31 +1006,142 @@ public bool IsSpawnProtected
+        /// <summary>
+        /// Gets all the current players of a specific role.
+        /// </summary>
+        internal static Dictionary<RoleType, HashSet<Player>> InstantiatedRolesToPlayersInternal { get; private set; } = new()
```
Oh, you made comments, I made changes, review again

IRacle1 replied:
yea, based

Jesus-QC commented:
I dont like those internals, they can be broken with updates easily.

Just make them a methods to generate them dynamicly.

Undid-Iridium commented:
> I dont like those internals, they can be broken with updates easily.
> 
> Just make them a methods to generate them dynamicly.

https://github.com/Exiled-Team/EXILED/pull/1300#discussion_r909194933

Undid-Iridium commented:
Closing due to inactivity from reviewers. If someone wants this, re-open or create new PR.
